frontend/catalyst/compiler.py

```diff
@@ -119,37 +121,115 @@ def get_lib_path(project, env):
     "--relocation-model=pic",
 ]
 
-linker = "c99"
 
-mlir_lib_path = get_lib_path("llvm", "MLIR_LIB_DIR")
-linker_common_flags = [
-    "-shared",
-    "-rdynamic",
-    f"-L{mlir_lib_path}",
-    "-Wl,-no-as-needed",
-    f"-Wl,-rpath,{mlir_lib_path}",
-]
+# pylint: disable=too-few-public-methods
+class AbstractLinker:
+    """Linker interface
 
-common_libs = [
-    "-lpthread",
-    "-lmlir_c_runner_utils",  # required for memref.copy
-]
+    In order to avoid relying on a single linker at run time and allow the user some flexibility,
+    this class defines a linker resolution order where multiple known linkers are attempted. The
+    order is defined as follows:
 
-runtime_libs = [
-    "-lrt_backend",
-    "-lrt_capi",
-    *common_libs,
-]
+    1. A user specified linker via the environment variable CATALYST_CC. It is expected that the
+        user provided linker is flag compatilble with LD.
```

Comment:
Shouldn't it be flag compatible with gcc? Since we don't use an actual linker but a compiler as a linker driver?

Reply:
The discrepancy here is just whether we consider GCC a linker or a linker driver. The underlying linker is LD. You are correct in stating that LD will need more flags because GCC adds default flags as options to LD. What I'm saying here is that the flags must be possible in LD in the first place. We cannot add "-some-flag-that-doesn't-exist-in-ld" as a flag. I'll just change it to GCC.

Reply:
Well we don't control what linker will be invoked, isn't that determined by the linker driver?

Reply:
In addition, the text should probably reflect that we expect a _compiler_ rather than a _linker_ for that environment variable. I'm assuming if the user were to actually pass in a linker the invocation would fail (e.g. I'm assuming `-Wl,` is not understood by the linker itself).

Reply:
> [...] do we pass flags that are specific to LD (I guess the -Wl, style flags)?

We haven't tested other linkers: gold, mold, perhaps others?. All I'm saying is that the linker in (driver, linker) needs to understand the flags we have. As a note, the driver also needs to understands the flags we pass, which the driver consumes. These flags are compatible with LD and LLD. But other linkers might not understand them. Or additionally, other drivers might have different behaviours (add different flags to the linker for example).

> Well we don't control what linker will be invoked, isn't that determined by the linker driver?

Yes. I believe compilers can be configured to invoke different linkers by default.  See for [example](https://llvm.org/docs/CMake.html#frequently-used-llvm-related-variables)

> LLVM_USE_LINKER:STRING
>    Override the system’s default linker. For instance use lld with -DLLVM_USE_LINKER=lld. 

> LLVM_USE_LINKER:STRING
>    Add -fuse-ld={name} to the link invocation. The possible value depend on your compiler, for clang the value can be an absolute path to your custom linker, otherwise clang will prefix the name with ld. and apply its usual search. For example to link LLVM with the Gold linker, cmake can be invoked with -DLLVM_USE_LINKER=gold.

If we wanted control over which linker is run, we would invoke the linker directly and add more flags. Perhaps we can just add `-fuse-ld=ld` to our current flags.

> In addition, the text should probably reflect that we expect a compiler rather than a linker for that environment variable. I'm assuming if the user were to actually pass in a linker the invocation would fail (e.g. I'm assuming -Wl, is not understood by the linker itself).

That's a good suggestion.